envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotsVersions.kt

```diff
@@ -22,23 +23,37 @@ internal class SnapshotsVersions {
 
     private val versions = ConcurrentHashMap<Group, VersionsWithData>()
 
-    fun version(group: Group, clusters: List<Cluster>, endpoints: List<ClusterLoadAssignment>): Version {
+    fun version(
+        group: Group,
+        clusters: List<Cluster>,
+        endpoints: List<ClusterLoadAssignment>,
+        listeners: List<Listener> = listOf()
+    ): Version {
         val versionsWithData = versions.compute(group) { _, previous ->
             val version = when (previous) {
-                null -> Version(clusters = ClustersVersion(newVersion()), endpoints = EndpointsVersion(newVersion()))
+                null -> Version(
+                        clusters = ClustersVersion(newVersion()),
+                        endpoints = EndpointsVersion(newVersion()),
+                        listeners = ListenersVersion(newVersion())
+                )
                 else -> {
                     val clustersChanged = previous.clusters != clusters
                     Version(
                         clusters = selectClusters(previous, clustersChanged),
-                        endpoints = selectEndpoints(previous, endpoints, clustersChanged)
+                        endpoints = selectEndpoints(previous, endpoints, clustersChanged),
+                        listeners = selectListeners(previous, previous.listeners != listeners)
```

Comment:
We need to send RDS after LDS: `When a listener is added, it will be “warmed” before taking traffic. For example, if the listener references an RDS configuration, that configuration will be resolved and fetched before the listener is moved to “active.”` https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/lds